File: frontend/src/types/execution.ts
export type ExecutionStatus = "not_started" | "in_progress" | "delayed" | "completed" | "cancelled";

export interface Execution {
  id: number;
  public_contract_id: number;
  avancement: number;
  date_debut?: string | null;
  date_fin_prevue?: string | null;
  date_fin_reelle?: string | null;
  statut: ExecutionStatus;
  observations?: string | null;
  created_at: string;
  updated_at: string;
}

export interface ExecutionCreate {
  public_contract_id: number;
  avancement?: number;
  date_debut?: string;
  date_fin_prevue?: string;
  date_fin_reelle?: string;
  observations?: string;
}

export type ExecutionUpdate = Partial<ExecutionCreate> & {
  statut?: ExecutionStatus;
};

export interface ExecutionReport {
  id: number;
  execution_id: number;
  title: string;
  description?: string | null;
  progress_percentage: number;
  report_file_url?: string | null;
  created_by_id: number;
  created_at: string;
}

export interface ExecutionReportCreate {
  execution_id: number;
  title: string;
  description?: string;
  progress_percentage: number;
  report_file_url?: string;
}

File: backend/app/models/execution.py
from sqlalchemy import Column, Integer, Float, String, DateTime, ForeignKey, Text, func
from sqlalchemy.orm import relationship

from ..database import Base


class Execution(Base):
    __tablename__ = "executions"

    id = Column(Integer, primary_key=True, index=True)
    public_contract_id = Column(Integer, ForeignKey("public_contracts.id"), nullable=False, unique=True)
    avancement = Column(Float, nullable=False, default=0.0)
    date_debut = Column(DateTime(timezone=True), nullable=True)
    date_fin_prevue = Column(DateTime(timezone=True), nullable=True)
    date_fin_reelle = Column(DateTime(timezone=True), nullable=True)
    statut = Column(String(50), nullable=False, default="not_started")
    observations = Column(Text, nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    public_contract = relationship("PublicContract", back_populates="execution")
    progress_reports = relationship("ExecutionReport", back_populates="execution")
